### PR TITLE
Fix CCE in Add Check Code action

### LINKED ISSUE #33167: Hover on spread field causes CCE in `AddCheckCodeAction`
**Description:**
Consider the below sample 
```bal
function init() {
    record {|int field1?; string field2;|} c = {field2: "Laptop"};
    record {|int field3; string...;|} var4 = {readonly field3: 45434, ...c};
}
```
Causes CCE in `AddCheckCodeAction` when hover `c`
<details>
  <summary>Stack trace.(Click to expand)</summary>

```
[Error - 1:55:11 PM] CodeAction 'AddCheckCodeAction' failed! {, error: 'class org.wso2.ballerinalang.compiler.util.Name cannot be cast to class io.ballerina.compiler.api.symbols.TypeSymbol (org.wso2.ballerinalang.compiler.util.Name and io.ballerina.compiler.api.symbols.TypeSymbol are in unnamed module of loader 'app')'}
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.util.Name cannot be cast to class io.ballerina.compiler.api.symbols.TypeSymbol (org.wso2.ballerinalang.compiler.util.Name and io.ballerina.compiler.api.symbols.TypeSymbol are in unnamed module of loader 'app')
	at org.ballerinalang.langserver.codeaction.providers.AddCheckCodeAction.getDiagBasedCodeActions(AddCheckCodeAction.java:76)
	at org.ballerinalang.langserver.codeaction.CodeActionRouter.lambda$getAvailableCodeActions$4(CodeActionRouter.java:107)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.ballerinalang.langserver.codeaction.CodeActionRouter.lambda$getAvailableCodeActions$5(CodeActionRouter.java:101)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.ballerinalang.langserver.codeaction.CodeActionRouter.getAvailableCodeActions(CodeActionRouter.java:99)
	at org.ballerinalang.langserver.codeaction.BallerinaCodeActionExtension.execute(BallerinaCodeActionExtension.java:46)
	at org.ballerinalang.langserver.codeaction.BallerinaCodeActionExtension.execute(BallerinaCodeActionExtension.java:34)
	at org.ballerinalang.langserver.LangExtensionDelegator.codeActions(LangExtensionDelegator.java:157)
	at org.ballerinalang.langserver.BallerinaTextDocumentService.lambda$codeAction$8(BallerinaTextDocumentService.java:328)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
```
</details>

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject by introducing a systematic way of accessing diagnostic properties of a given diagnostic. 

Fixes #33167

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java [hunk 1]
```diff
@@ -67,8 +68,14 @@ public AddCheckCodeAction() {
...
+                positionDetails.getPropertyIndex(diagnostic.diagnosticInfo().code(),
+                        DiagnosticPropertyKey.DIAG_PROP_INCOMPATIBLE_TYPES_FOUND);
+        if (propertyIndex.isEmpty()) {
+            return Collections.emptyList();
+        }
+        
+        Optional<TypeSymbol> foundType = positionDetails.diagnosticProperty(propertyIndex.get());
         if (foundType.isEmpty()) {
             return Collections.emptyList();
         }
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java:81]: These lines seem to be duplicated, shall we fix that?

**Review by gayalkuruppu — COMMENTED** [review]
Changes look good

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/spi/DiagBasedPositionDetails.java [hunk 2]
```diff
@@ -51,4 +53,14 @@
      * @return Value of the property at given index
      */
     <T> Optional<T> diagnosticProperty(int propertyIndex);
+
+    /**
+     * Returns property index given the code.
+     * 
+     * @param diagnosticCode Diagnostic code.
+     * @param propertyName Name of the property.
+     * @return
```
Comment (gayalkuruppu) [inline language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/spi/DiagBasedPositionDetails.java:62]: Shall we complete the @return line? We can may be change the line 58 to `Get property index given the code.`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddCheckCodeAction.java]: Done

File: language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/spi/DiagBasedPositionDetails.java (hunk 2 above)
Comment (malinthar) [reply language-server/modules/langserver-commons/src/main/java/org/ballerinalang/langserver/commons/codeaction/spi/DiagBasedPositionDetails.java]: Done

**Review by nadeeshaan — CHANGES_REQUESTED** [review]
With this itself, we need to avoid the following as well.


https://user-images.githubusercontent.com/1329674/141980523-c3e81e5b-4f0f-4715-bd79-fc7519f9a5c2.mp4

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractCodeActionProvider.java [hunk 3]
```diff
@@ -125,4 +127,49 @@ public static CodeAction createQuickFixCodeAction(String commandTitle, List<Text
...
+                    return Optional.of(2);
+                }
+                break;
+            case DIAG_PROP_INCOMPATIBLE_TYPES_EXPECTED:
+                if ("BCE2066".equals(diagnosticCode) || "BCE2068".equals(diagnosticCode)) {
+                    return Optional.of(0);
+                }
+                break;
+            default:
+                return Optional.empty();
```
[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractCodeActionProvider.java]: I have forgot to remove this

[the remark this answers is not in the document]
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AbstractCodeActionProvider.java]: I have already removed this

**Comment by nadeeshaan** [comment]
> > With this itself, we need to avoid the following as well.
> > Nov-16-2021.17-21-02.mp4
> 
> @nadeeshaan I have created a separate issue for this. This seems to be a general issue with the `getMatchedNode()` and `getExpression()` method. The relevant issues are #33792 #33783 #33780
> 
> Thought that the above issues are not in the scope of this PR. However, can fix them up if needed. WDYT?

If we have separate issues, let's address them with that

**Review by nadeeshaan — APPROVED** [review]

**Review by gayalkuruppu — APPROVED** [review]
LGTM

**Pull request merged**